---
title: "Gene Coexpression Networks"
output: html_document
---

## Problem approach and initial preparations

In order to do this assignment, a number of custom functions were created. This are located in the file named "EnrichmentAnnotationFunctions.R". There are three functions, whose description is as follows:

*Celltype_annotator*: It takes two input parameters, a data.frame with the annotations generated by *CoExpNets::cellTypeByModule* and the name of the module of interest. It returns a data.frame with the celltypes associated with the given module.

*Functional_annotator*: Similarly to the previous one, it takes in a data.frame with the annotations generated by *CoExpNets::getGProfilerOnNet* and a module name butreturns a list of data.frames with the top 10 most significant annotations based on the p-values, one for each kind of annotation (BP, MF, CC, Kegg, Reactome).

*Topmm_genextractor*: It takes as input parameters a data.frame generated by the function *getMM* applied to all genes in the dataset, the name of the module and a cutoff value to control how many genes are returned. The output consists of a data.frame with as many genes as specified in cutoff and the gene names in both Ensembl and GeneName encoding as well as the actual values of Module Membership of the genes

Appart from these functions, only the packages CoExpNets, CoExpROSMAP and WGCNA were used.

## Library loading

Now the mentioned packages were loaded as well as the file with the custom functions.

```{r}
####Change to desired working directory####
setwd("/home/albert/Bioinfo/MachineLearning/TareasFinales")
library(CoExpNets)
library(CoExpROSMAP)
library(WGCNA)
source("EnrichmentAnnotationFunctions.R")
```

## Data acquisition and visualization
First, the CoExpROSMAP database was initialized :

```{r}
CoExpROSMAP::initDb()
```

Now we verified which data are available in it

```{r}
for(data.family in getNetworkCategories()){
  cat(paste0("Family of gene expression dataset ",data.family," available\n"))#Data family name
  tissues = getAvailableNetworks(data.family)
  for(tissue in tissues){#Names of the datasets. Ad means Alzheimer disease"
    cat(paste0("Gene expression profiling dataset for tissue ",tissue,"\n is ",
               getExprDataFromTissue(which.one=data.family,tissue=tissue,only.file = T)," available\n"))
  }
}
```

In this assignment only the last dataset, called allsamples was used. This one includes subjects with and without Alzheimer disease. Every dataset has already been preprocessed, as can be inferred from the directory names which includes FPKM, a normalization technique applied to transcriptomic data.

```{r}
expr_data<-getExprDataFromTissue(tissue = "allsamples", which.one = "CoExpROSMAP")
dim(expr_data)
```

Next comes the retrieval of the metadata, which includes several covariables, some technical such as post mortem interval (pmi) and other with biological significance to the Alzheimer disease.

```{r}
metadata<-CoExpROSMAP::getCovariates(tissue="allsamples", which.one="CoExpROSMAP")
```

The covariable braaksc is the one that describes the state of progression of the disease. In a simple barplot we visualized the distribution of the values of braaksc, which there are 7 of, across the individuals from every data available in the databse. 

```{r}
nets<-getAvailableNetworks("CoExpROSMAP")
par(mfrow=c(2,2))
devnull<-lapply(nets, function(x){
  covs<-CoExpROSMAP::getCovariates(tissue=x, which.one = "CoExpROSMAP")
  barplot(table(covs$braaksc),main=paste("Braak stage in", x, sep = " "))
  return(covs)
})
par(mfrow=c(1,1))
```
The levels 4 and 5 are the most prominent in all data set, which means that these values are not particulary correlated with the disease even though they are the most frequent values in both ad and probad dataset. This covariable alone is not sufficient to classify the individuals, more information is needed, such as the one provided by gene expression.

## Data filtering

In this project only the individuals with a braaksc value equal to 5 or 6 were included in the analysis. So we selected the appropiate rows of the dataset as following: 

```{r}
rowstoinclude<-rownames(metadata)[c(which(metadata$braaksc==5), which(metadata$braaksc==6))]
expr_data<-expr_data[rownames(expr_data) %in% rowstoinclude,]
```
## Network creation 

Now that the data is ready, we used the function *CoExpNets::getDownstreamNetwork* to create the coexpression network. The function uses unsupervised learning alogorithms to create modules of genes with similar expression patterns. Since the computation is time-consuming, we allowed the possibilty to load the .rds object with the network if it has already been created once, to avoid executing this function. All the results of the analysis that were be performed were saved in a directory named CoexpressionResults, which is created in the current working directory.

```{r}
#####Change to FALSE if first time###################
loadnetwork<-TRUE                                   #
#####################################################
if(loadnetwork==FALSE){#Avoid executing twice 
  system("mkdir CoexpressionResults")
  expr_network<-CoExpNets::getDownstreamNetwork(tissue = "MyRosMap",
                                              n.iterations = 20,
                                              net.type = "signed",
                                              debug = FALSE,
                                              expr.data = expr_data,
                                              job.path = "./CoexpressionResults")
}else{
  expr_network<-readRDS("./CoexpressionResults/netMyRosMap.13.it.20.rds")
}
networkname<-"./CoexpressionResults/netMyRosMap.13.it.20.rds"
```
Before continuing with the annotation of the network, a small formatting needs to be done. Some genes are encoded with their Ensembl accesion number followed by a '.' and another number. These last two elements must be removed in order to succesfully apply the annotation functions to the network object.
```{r}
names(expr_network$moduleColors)<-gsub(pattern = "\\.\\d+", replacement = "", 
                                       perl = TRUE, names(expr_network$moduleColors))
```
## Network annotation

Three types of annotations of the network's modules were performed. Firstly, a functional annotation of Gene Ontology terms, including all three of them, that is biokogical process (BP), molecular function (MF) and cellular component (CC). Additionally, Reactome and Kegg annotations were obtained. Afterwards, a cellular type enrichment was computed and finally, the correlation and with the covariables of the metadata and its significance was calculated.

### Functional annotation

The function *CoExpNets::getGProfilerOnNet* was used. It generates a data.frame with all the annotations found for each module, its type and statistical significance. The function *CoExpNets::getDownstreamNetwork* used beforehand also generates as default a similar output, but it is not identical to the one created here. Hence, to distinguish them , this output file ends with *"gprofII"*. The function make use of the package gprofiler to obtain the annotation terms.

```{r}
go_data<-CoExpNets::getGProfilerOnNet(net.file=networkname, exclude.iea=FALSE, out.file=paste(
  networkname, "gprofII.csv", sep = "_"))
```
### Cellular type enrichment

The function *CoExpNets::cellTypeByModule* was employed this time. It generates a .pdf file with the enrichment plot, where it can be seen which cellular type are the modules enriched in and the significance of the association, the more intense the color, the more significant it is. It also returns an object of type matrix which contains the information used to produce the aforementioned plot. 
In a similar fashion to the previously done functional annotaion, the *CoExpNets::getDownstreamNetwork* creates by default a .csv file with the information of the enrichment, but it is also not identical to the one computed here. Therefore, this one was saved directly as "CelltypeCorrelation.csv" to the output directory

```{r}
CoExpNets::initDb()
cells_data<-CoExpNets::cellTypeByModule(return.processed = FALSE,
                                        tissue="MyRosMap",
                                        which.one = "new",
                                        plot.file=paste(networkname,".celltype.pdf", sep = "_"),
                                        net.in=networkname,
                                        legend="ROS/MAP cell type signals")
cells_data<-as.data.frame(cells_data)#Now it can be worked with normally
write.csv(cells_data, "./CoexpressionResults/CelltypeCorrelation.csv", quote = FALSE)
```
### Correlation significance with the covariables

This time, the function *CoExpNets::corWithCatTraits* was utilized. It requires as input that the covariables, which are stored as a data.frame in metadata, are ordered identically as the samples are in the expression data used to obtain the coexpression network. The output of the function includes a .pdf file with a heatmap plot showing the significance of the correlation as -log10(p-value). The data.frame that contains the values represented in the plot was also saved as a .csv file

```{r}
filt_metadata<-metadata[rowstoinclude, ]
m<-match(rownames(expr_data), rownames(filt_metadata))
filt_metadata<-filt_metadata[m,]
stopifnot(identical(rownames(expr_data),rownames(filt_metadata)))#Make sure they are identical
covarCorr_data<-CoExpNets::corWithCatTraits(which.one="new", tissue=networkname,
                                            covs=filt_metadata,
                                            covlist = colnames(metadata))
write.csv(covarCorr_data, file = "./CoexpressionResults/CovariableCorrelation.csv", quote = FALSE)
```

## Module visualization

Prior to the analysis of the modules, a quick visualization of some of their features was performed. This included their size, the correlation between them and their hierarchical relationships.

To plot the size, the function *CoExpNets::plotModSizes* is used. The size is represented as a barplot. The height of the bars corresponds to the number of gene that consititutes each module

```{r}
CoExpNets::plotModSizes(tissue=networkname,which.one="new")
```
To see the correlation of the modules, first a so called Eigengene of each module were calculated, using the function *getNetworkEigengenes*. This is not a true gene, rather is a variable constructed by performing PCA with the gene expression values of the modules. To be precise, is the first component of the PCA, and therefore is a "theoretical" gene that represents the best the whole module.
First we showd the correlation with a corrplot, and then a hierarchical cluster as a dendogram, using *plotEGClustering*

```{r}
library(corrplot)
egs<-getNetworkEigengenes(which.one="new",tissue=networkname)
corrplot(cor(egs), type = "upper", tl.cex = 0.5, order='hclust')
plotEGClustering(which.one="new",tissue=networkname)
```
The correlated modules appear close together in the dendogram.

La correlación de módulos con las variables deja ver una clara relación entre los módulos y el intervalo post-mortem, que es una variable técnica, que no tiene relación con la parte biológica. Es significancia de correlación. A partir de 1.3 es significativo, nosotros lo hemos puesto en 2.

## Seleccion de modulos

Basados en correlacion con covariables. 
```{r}
selectModules<-as.character(sapply(1:nrow(covarCorr_data), function(x){
  if(length(which((covarCorr_data[x,]>2))) > 0){ #Coge los modulos que tengan corr > 2 en CUALQUIER covariable
    return(rownames(covarCorr_data)[x])
  }
}))
selectModules<-selectModules[selectModules!="NULL"]
```


Basados en numero de terminos de anotación
Se escogen los 4 módulos que tengan un mayor número de ¿términos anotados?
```{r}
term_matrix<-as.matrix(sort(table(go_data$query.number), decreasing = TRUE))
selectModulesII<-as.character(rownames(term_matrix)[1:4])
```

Se guardan juntos todos los módulos a analizar y los distintos tipos de anotaciones. 
```{r}
totalmod<-c(selectModules, selectModulesII)
annotations <- c("CC","MF","BP","keg","rea", "celltype")
```

## Analisis de los modulos 

Se hace un bucle iterando por cada módulo y por cada anotación para obtener todas las anotaciones de todos los módulos. 
```{r}
annot_list <- lapply(totalmod, function(x){
  aux_list <- lapply(annotations,function(y){
    if(y != "celltype"){
      return(functional_annotator(data=go_data, modulename = x, annot_type = y))
    }else{
      return(celltype_annotator(data = cells_data, modulename = x))
    }
  })
  names(aux_list) <- annotations
  return(aux_list)
  })
names(annot_list) <- totalmod
```

Se obtiene una lista con todos los módulos, que a su vez cada uno tiene 6 listas que son cada una de las anotaciones. 
Se procede a la extracción de los genes con el MM más alto de cada módulo para su análisis.
```{r}
mm<-getMM(tissue = networkname, which.one = "new", genes = colnames(expr_data),
          expr.data.file = expr_data) #Extrae el mm de todos los genes
TopGenes_list<-lapply(totalmod, function(x){
  df<-topmm_genextractor(mmdata = mm, modulename = x, cutoff = 5)
  return(df)
})
names(TopGenes_list)<-totalmod
```


## Descripción de los módulos

### Thistle1

The Thistle1 module was made up of genes related to the mitochondria, specifically to the internal membrane and the protein complexes found there. Functionally they were related with electro transport and ATP synthesis. They were also related with neurodegenerative diseases such as Alzheimer's, Parkinson´s and Huntinton´s. Theses disease were related with alterations in the function of mitochondrias that were related with energy production, apoptosis and ROS (Reactive Oxygen Species) production, although in none of them it was completely known the molecular mechanism that originated them (Moreira _et al_, 2010). At cellular level, genes of this module were associated with neurons, which are the most important cell type in Alzheimer's disease.

The gen with highest MM (Module Membership) is APOO, which encodes apolipoprotein. This protein is associated with the formation of contact points between mitochondria and the structural organization of mitochondrial ridges. 
The other four genes with highest MM are SLC30A5, a zinc transporter; UPRT, an uracil-phosphoribosyl transferase; LEO1,  a component of the RNA polymerase II-dependent transcription activator complex; and IFTAP, a protein associated with intraflagellar transport. The functions of these genes are not closely related to the other annotations in the module.

### Royalblue

The royalblue module genes related to the membrane and membranous organelles. Their function was that of receptors, which binded to various molecules such as extracellular matrix proteins, cytokines and peptides such as TNF (Tumor Necrosis Factor). Consequently they were involved in several signalling pathways that depended on some membrane receptor to detect a signal. 
It appeared in response to cytiknes, other interleukines, TNF, HIF (Hypoxia Inducible Factor), cytotoxicity caused by the action of NK cells. Neutrophil degranulation was related with immune system. These pathways triggered among other responses apoptosis and inflammation. Nerve cell apoptosis and inflammation  occured in the course of neurodegenerative diseases, including Alzheimer's. It was also interesting that the enrichment found in the microglial cell type, since these cells were of myeloid origin. That is, they came from macrophages, cells with phagocytic and antigen-presenting activity among others. Within the nervous tissue, they differentiated into microglia and fullfilled immune functions, phagocyting foreign bodies such as apoptotic cells and amyloid deposits that were formed in Alzheimer's disease. There are studies that relate microglial cells to the origin of the disease (Hemonnot _et al_. 2018).

The five genes with the highest MM of royalblue module are HYAL2, a liposomal hialuronidase enzyme, ICAM2, a intercellular binding protein, DGKH, a diacylglycerol kinase induced by glicocorticoids, CPD, a carboxipeptidase enzyme, and SNAT1, a transcriptional repressor. 
The most relevant functions in relation to the previously commented notes are that of ICAM2, which participates in the binding between cells,that forms part of the functions of microglial cells, and DGKH, because it regulates the levels of a second messenger such as diacylglycerol, which is involved in several signalling pathways.

### Tan

The tan module contained genes that act as transcription factors... They were related with anabolic reactions and macromolecules biosynthesis. They appeared associated with some signalling pathways such as MAP kinases and TFG (Tumor Growth Factor). One the one hand, there were studies that established a link between deficiency in TGF dependent signalling and Alzheimer's disease (Hu _et al_., 2019). On the other hand, there were genes associated with various types of cancer, such as lung, colorrectal, breast ant thyroid cancer; as well as the human leukemia virus, which it was involved  in cancer due to its replication cycle. There were studies about an inverse correlation between cancer and Alzheimer's disease, since cancer is caused by an uncontrolled proliferation of cells and Alzheimer's by the opposite cause, an increase in cell death (Anne Feng _et al_. , 2017). Again, the cell type associated with the genes of this module the microglia.

The three genes with the highest MM are TBC1D10, an Rab type GTPase-activating protein involved in the regulation of vesicle transport in oligodendrocytes; OXSR1, a kinase that is activated in response to oxidative stress; and a non-traduced transcript, a non-coding RNA called LINC01354 with no attributed function.  Of these three, OXSR1 is perhaps the most relevant because of its relationship to oxidative stress and its role in apoptosis.

### Darkturquoise

The darkturquoise module had genes related to signalling pathways, immune response to virial infections through pathways such as interferon and proteins that interact with RNA. Its genes were expressed in the cellular cytosol. Cell types associated with this module were T helper lymphocytes and the microglia, both related with immune system. The fact that some  virus types may be related to the neuropathology of alzheimer's disease it is interested (Sochocka _et al_ 2017).

CXCL5 is a cytokine from the CXC family. SAMD9L is a gen whose alteration involves diseases that affect the brain such as ataxia pancytopenia syndrome. UBA7 is involved in ubiquitination. PATL2 participates in RNA binding, and its alteration is related to fertility problems. 

### Purple

The purple module was enriched exclusively in mitochondria, with no associated cell types, focusing on metabolism and activity as a cofactor and oxidoreductase. Interestingly, it was involved in a multitude of metabolic processes, including basal metabolism and fatty acid metabolism. A certain association was found between fatty acid metabolism and Alzheimer's disease (Snowden _et al_ 2017 ).

LINC00526 encodes for a long non-coding RNA. C1orf194 is a little studied gen and it would be the cause for different brain diseases. ZNF706 is a transcription factor involved in the negative regulation  of embryonic mesenchymal cell population maintenance, in addition to being involved in the negative regulation of transcription. RP4-612B15.3 is a pseudogen presents in immune system populations. TOE1 inhibits cell growth by stopping the cell cycle. 

### Floralwhite

The floralwhite module was related to the immune system at the level of intracellular (receptor signalling) and extracellular (immunoglobulin G (Ig G) signaling, as well as activation of myeloid leukocytes. It had been found that during the development of Alzheimer's disease, there was expression of proinflammatory molecules in myeloid cells (Thome _et al_ 2018). In this module a relation was appreciated at the level of disease and cellular type, since it was enriched in rheumatoid arthritis and in differentiation of osteoclasts, and it had been seen that an excessive activity of the osteoclasts was related to the appearance of rheumatoid arthritis (Nakashima _et al_ 2008). No relationship was found between rheumatoid arthritis and Alzheimer's disease (Qixuan _et al_ 2018), so it would be possible that they only shared this pro-inflammatory profile and that is why it was highlighted in the analysis.
 
CPED1 is a gen of which there is no known function. PDCD4-AS1 encodes a long non-coding RNA. It is associated with breast cancer. C11orf45 encodes an uncharacterized protein. AC018462.3 encodes a long non-coding RNA. EMB is related with MAP kinase signalling pathway, and the transport of glucose and other sugar molecules. 



# Bibliografía

Paula I. Moreira , Cristina Carvalho, Xiongwei Zhu,  Mark A. Smith , George Perry : Mitochondrial dysfunction is a trigger of Alzheimer's disease pathophysiology. _Biochimica et Biophysica_ Acta 2010, 1802 2:10

Anne-Laure Hemonnot , Jennifer Hua,  Lauriane Ulmann, and Hélène Hirbec Microglia in Alzheimer Disease: Well-Known Targets and New Opportunities. _Frontiers in Aging Neuroscience_ 2019, 11 233

Yueqiang Hu, Wei Chen, Lin Wu,  Lingfei Jiang , Ni Liang , Lulu Tan , Minghui Liang, Nong Tang: TGF-β1 Restores Hippocampal Synaptic Plasticity and Memory in Alzheimer Model via the PI3K/Akt/Wnt/β-Catenin Signaling Pathway. _Journal of Molecular Neuroscience_ 2019, 67, 142:149. 

Yen-Chen Anne Feng,  Kelly Cho, Sara Lindstrom,  Peter Kraft, Jean Cormack: Investigating the genetic relationship between Alzheimer’s disease and cancer using GWAS summary statistics. _Human Genetics_ 2017, 136, 1341:1351.

Marta Sochocka, Katarzyna Zwolińska, Jerzy Leszek. The Infectious Etiology of Alzheimer’s Disease. _Curr Neuropharmacol_ 2017, 15(7): 996–1009.

Stuart G. Snowden, Amera A. Ebshiana, Abdul Hye, Yang An, Olga Pletnikova, Richard O’Brien, John Troncoso, Cristina Legido-Quigley, Madhav Thambisetty. Association between fatty acid metabolism in the brain and Alzheimer disease neuropathology and cognitive performance: A nontargeted metabolomic study. _PLOS Medicine_ 2017, Mar 21;14(3):e1002266.

Aaron D. Thome, Alireza Faridar, David R. Beers, Jason R. Thonhoff, Weihua Zhao, Shixiang Wen, Belen Pascual, Joseph C. Masdeu,Stanley H. Appel. Functional alterations of myeloid cells during the course of Alzheimer’s disease.  _Molecular Neurodegeneration_ 2018, 13:61 

Qixuan Cai, Zhuoyuan Xin, Lin Zuo, Fan Li, Bin Liu. Alzheimer’s Disease and Rheumatoid Arthritis: A Mendelian Randomization Study.  _Frontiers in Neuroscience_ 2018, 12, 627.
