---
title: "Gene Coexpression Networks"
output: html_document
---

## Problem approach and initial preparations

In order to do this assignment, a number of custom functions were created. These are located in the file named "EnrichmentAnnotationFunctions.R". There are three functions, whose description is as follows:

*Celltype_annotator*: It takes two input parameters, a data.frame with the annotations generated by *CoExpNets::cellTypeByModule* and the name of the module of interest. It returns a data.frame with the celltypes associated with the given module.

*Functional_annotator*: Similarly to the previous one, it takes in a data.frame with the annotations generated by *CoExpNets::getGProfilerOnNet* and a module name but returns a list of data.frames with the top 10 most significant annotations based on the p-values, one for each kind of annotation (BP, MF, CC, Kegg, Reactome).

*Topmm_genextractor*: It takes as input parameters a data.frame generated by the function *getMM* applied to all genes in the dataset, the name of the module and a cutoff value to control how many genes are returned. The output consists of a data.frame with as many genes as specified in cutoff and the gene names in both Ensembl and GeneName encoding as well as the actual values of Module Membership of the genes.

Appart from these functions, only the packages CoExpNets, CoExpROSMAP and WGCNA were used.

## Library loading

Now the mentioned packages were loaded as well as the file with the custom functions.

```{r}
####Change to desired working directory####
setwd("/home/albert/Bioinfo/MachineLearning/TareasFinales")
library(CoExpNets)
library(CoExpROSMAP)
library(WGCNA)
source("EnrichmentAnnotationFunctions.R")
```

## Data acquisition and visualization

First, the CoExpROSMAP database was initialized :

```{r}
CoExpROSMAP::initDb()
```

Now the available data in it was displayed and verified:

```{r}
for(data.family in getNetworkCategories()){
  cat(paste0("Family of gene expression dataset ",data.family," available\n"))#Data family name
  tissues = getAvailableNetworks(data.family)
  for(tissue in tissues){#Names of the datasets. Ad means Alzheimer disease"
    cat(paste0("Gene expression profiling dataset for tissue ",tissue,"\n is ",
               getExprDataFromTissue(which.one=data.family,tissue=tissue,only.file = T)," available\n"))
  }
}
```

In this assignment only the last dataset, called all samples was used. This one includes subjects with and without Alzheimer´s disease. Every dataset has already been preprocessed, as can be inferred from the directory names which includes FPKM, a normalization technique applied to transcriptomic data.

```{r}
expr_data<-getExprDataFromTissue(tissue = "allsamples", which.one = "CoExpROSMAP")
dim(expr_data)
```

Next comes the retrieval of the metadata, which includes several covariables, some technical such as post mortem interval (pmi) and other with biological significance to the Alzheimer´s disease.

```{r}
metadata<-CoExpROSMAP::getCovariates(tissue="allsamples", which.one="CoExpROSMAP")
```

The covariable braaksc is the one that describes the state of progression of the disease. In a simple barplot, the distribution of the values of braaksc was visualized, which there are 7 o; across the individuals from every data available in the database. 

```{r}
nets<-getAvailableNetworks("CoExpROSMAP")
par(mfrow=c(2,2))
devnull<-lapply(nets, function(x){
  covs<-CoExpROSMAP::getCovariates(tissue=x, which.one = "CoExpROSMAP")
  barplot(table(covs$braaksc),main=paste("Braak stage in", x, sep = " "))
  return(covs)
})
par(mfrow=c(1,1))
```
The levels 4 and 5 are the most prominent in all data set, which means that these values are not particulary correlated with the disease even though they are the most frequent values in both ad and probad dataset. This covariable alone is not sufficient to classify the individuals, more information is needed, such as the one provided by gene expression.

## Data filtering

In this project only the individuals with a braaksc value equal to 5 or 6 were included in the analysis. So  the appropiate rows of the dataset were selected as following: 

```{r}
rowstoinclude<-rownames(metadata)[c(which(metadata$braaksc==5), which(metadata$braaksc==6))]
expr_data<-expr_data[rownames(expr_data) %in% rowstoinclude,]
```
## Network creation 

Now that the data is ready, the function *CoExpNets::getDownstreamNetwork* was used to create the coexpression network. This  function uses unsupervised learning alogorithms to create modules of genes with similar expression patterns. Since the computation is time-consuming, the possibilty to load the .rds object with the network if it has already been created once, was allowed. All the results of the analysis that were be performed were saved in a directory named CoexpressionResults, which is created in the current working directory.

```{r}
#####Change to FALSE if first time###################
loadnetwork<-TRUE                                   #
#####################################################
if(loadnetwork==FALSE){#Avoid executing twice 
  system("mkdir CoexpressionResults")
  expr_network<-CoExpNets::getDownstreamNetwork(tissue = "MyRosMap",
                                              n.iterations = 20,
                                              net.type = "signed",
                                              debug = FALSE,
                                              expr.data = expr_data,
                                              job.path = "./CoexpressionResults")
}else{
  expr_network<-readRDS("./CoexpressionResults/netMyRosMap.13.it.20.rds")
}
networkname<-"./CoexpressionResults/netMyRosMap.13.it.20.rds"
```
Before continuing with the annotation of the network, a small formatting needs to be done. Some genes are encoded with their Ensembl accesion number followed by a '.' and another number. These last two elements must be removed in order to succesfully apply the annotation functions to the network object.
```{r}
names(expr_network$moduleColors)<-gsub(pattern = "\\.\\d+", replacement = "", 
                                       perl = TRUE, names(expr_network$moduleColors))
```
## Network annotation

Three types of annotations of the network's modules were performed. Firstly, a functional annotation of Gene Ontology terms, including all three of them, that is biological process (BP), molecular function (MF) and cellular component (CC). Additionally, Reactome and Kegg annotations were obtained. Afterwards, a cellular type enrichment was computed and finally, the correlation and with the covariables of the metadata and its significance was calculated.

### Functional annotation

The function *CoExpNets::getGProfilerOnNet* was used. It generates a data.frame with all the annotations found for each module, its type and statistical significance. The function *CoExpNets::getDownstreamNetwork* used beforehand also generates as default a similar output, but it is not identical to the one created here. Hence, to distinguish them , this output file ends with *"gprofII"*. The function make use of the package gprofiler to obtain the annotation terms.

```{r}
go_data<-CoExpNets::getGProfilerOnNet(net.file=networkname, exclude.iea=FALSE, out.file=paste(
  networkname, "gprofII.csv", sep = "_"))
```
### Cellular type enrichment

The function *CoExpNets::cellTypeByModule* was employed this time. It generates a .pdf file with the enrichment plot, where it can be seen which cellular type are the modules enriched in and the significance of the association, the more intense the color, the more significant it is. It also returns an object of type matrix which contains the information used to produce the aforementioned plot. 
In a similar fashion to the previously done functional annotaion, the *CoExpNets::getDownstreamNetwork* creates by default a .csv file with the information of the enrichment, but it is also not identical to the one computed here. Therefore, this one was saved directly as "CelltypeCorrelation.csv" to the output directory.

```{r}
CoExpNets::initDb()
cells_data<-CoExpNets::cellTypeByModule(return.processed = FALSE,
                                        tissue="MyRosMap",
                                        which.one = "new",
                                        plot.file=paste(networkname,".celltype.pdf", sep = "_"),
                                        net.in=networkname,
                                        legend="ROS/MAP cell type signals")
cells_data<-as.data.frame(cells_data)#Now it can be worked with normally
write.csv(cells_data, "./CoexpressionResults/CelltypeCorrelation.csv", quote = FALSE)
```
### Correlation significance with the covariables

This time, the function *CoExpNets::corWithCatTraits* was utilized. It requires as input that the covariables, which are stored as a data.frame in metadata, are ordered identically as the samples are in the expression data used to obtain the coexpression network. The output of the function includes a .pdf file with a heatmap plot showing the significance of the correlation as -log10(p-value). The data.frame that contains the values represented in the plot was also saved as a .csv file.

```{r}
filt_metadata<-metadata[rowstoinclude, ]
m<-match(rownames(expr_data), rownames(filt_metadata))
filt_metadata<-filt_metadata[m,]
stopifnot(identical(rownames(expr_data),rownames(filt_metadata)))#Make sure they are identical
covarCorr_data<-CoExpNets::corWithCatTraits(which.one="new", tissue=networkname,
                                            covs=filt_metadata,
                                            covlist = colnames(metadata))
write.csv(covarCorr_data, file = "./CoexpressionResults/CovariableCorrelation.csv", quote = FALSE)
```
The correlation of modules with the variables showed a clear relationship between the modules and the post-mortem interval, which is a technical variable, not related to the biological part of the analysis. 

## Module visualization

Prior to the analysis of the modules, a quick visualization of some of their features was performed. This included their size, the correlation between them and their hierarchical relationships.

To plot the size, the function *CoExpNets::plotModSizes* was used. The size is represented as a barplot. The height of the bars corresponds to the number of genes that consititutes each module.

```{r}
CoExpNets::plotModSizes(tissue=networkname,which.one="new")
```
To see the correlation of the modules, first a so called Eigengene of each module was calculated, using the function *getNetworkEigengenes*. This is not a true gene, rather is a variable constructed by performing PCA with the gene expression values of the modules. To be precise, it is the first component of the PCA, and therefore is a "theoretical" gene that represents the best the whole module.
First the correlation with a corrplot was showed, and then a hierarchical clustering as a dendogram, using *plotEGClustering*.

```{r}
library(corrplot)
egs<-getNetworkEigengenes(which.one="new",tissue=networkname)
corrplot(cor(egs), type = "upper", tl.cex = 0.5, order='hclust')
plotEGClustering(which.one="new",tissue=networkname)
```
The correlated modules appear closer together in the dendogram.

## Module selection

### Based on correlation with covariates 

A cutoff value of 2 was assigned to determine if a correlation is significant. Any module with a higher value for any correlation was chosen

```{r}
selectModules<-as.character(sapply(1:nrow(covarCorr_data), function(x){
  if(length(which((covarCorr_data[x,]>2))) > 0){ #Coge los modulos que tengan corr > 2 en CUALQUIER covariable
    return(rownames(covarCorr_data)[x])
  }
}))
selectModules<-selectModules[selectModules!="NULL"]
```
The chosen modules were Darkturquoise and Purple

### Based on number of annotation terms

The 4 modules with the highest query number were chosen, *i.e.* the ones with most annotation terms
```{r}
term_matrix<-as.matrix(sort(table(go_data$query.number), decreasing = TRUE))
selectModulesII<-as.character(rownames(term_matrix)[1:4])
```
These modules were Thistle1, Royalblue and Tan

All the modules that were going to be analysed and the different types of annotations were stored together in two vector variables. 
```{r}
totalmod<-c(selectModules, selectModulesII)
annotations <- c("CC","MF","BP","keg","rea", "celltype")
```

## Module analysis

A loop was made by iterating over the modules' names and over the types of annotation to obtain all the annotations of all the modules. The custom functions *functional_annotator* and *celltype_annotator* were employed.

```{r}
annot_list <- lapply(totalmod, function(x){
  aux_list <- lapply(annotations,function(y){
    if(y != "celltype"){
      return(functional_annotator(data=go_data, modulename = x, annot_type = y))
    }else{
      return(celltype_annotator(data = cells_data, modulename = x))
    }
  })
  names(aux_list) <- annotations
  return(aux_list)
  })
names(annot_list) <- totalmod
```
The output is a list of lists, one for every module, which in turn have 6 data.frames corresponding to each of the annotation types. 

Finally, to complement the analysis, the genes with the higher Module Membership of each one of the selected modules were extracted. The function *getMM* in combination with the custom *topmm_geneextractor* were utilized.

```{r}
mm<-getMM(tissue = networkname, which.one = "new", genes = colnames(expr_data),
          expr.data.file = expr_data) #Extrae el mm de todos los genes
TopGenes_list<-lapply(totalmod, function(x){
  df<-topmm_genextractor(mmdata = mm, modulename = x, cutoff = 5)
  return(df)
})
names(TopGenes_list)<-totalmod
```

## Module Description
Now for the last part, using the information recovered in the last section as well as relevant papers, a description of the modules in biological terms and their possible relation with the Alzheimer disease is presented below.

### Thistle1

The Thistle1 module was made up of genes related to the mitochondria, specifically to the internal membrane and the protein complexes found there. Functionally they were related with the electron transport and ATP synthesis. They are also related with neurodegenerative diseases such as Alzheimer's, Parkinson´s and Huntington´s. These disease are related with alterations in the function of mitochondrias that are related with energy production, apoptosis and ROS (Reactive Oxygen Species) production, although in none of them the molecular mechanism that originates it is completely understood (Moreira _et al_, 2010). At the cellular level, genes of this module were associated with neurons, which are the most important cell type in Alzheimer's disease.

The gen with highest MM (Module Membership) is APOO, which encodes apolipoprotein. This protein is associated with the formation of contact points between mitochondria and the structural organization of mitochondrial ridges. 
The other four genes with highest MM are SLC30A5, a zinc transporter; UPRT, an uracil-phosphoribosyl transferase; LEO1,  a component of the RNA polymerase II-dependent transcription activator complex; and IFTAP, a protein associated with intraflagellar transport. The functions of these genes are not closely related to the other annotations in the module.

### Royalblue

The Royalblue module genes were related to the membrane and membranous organelles. Their function is that of receptors, which bind to various molecules such as extracellular matrix proteins, cytokines and peptides such as TNF (Tumor Necrosis Factor). Consequently, they were involved in several signalling pathways that depended on some membrane receptor to detect a signal. 
Some of these pathways were response to cytokines, other interleukines, TNF, HIF (Hypoxia Inducible Factor) or cytotoxicity caused by the action of NK cells. Other process that was found was neutrophil degranulation, which is also related to the immune system. These pathways trigger, among other responses, apoptosis and inflammation. Nerve cell apoptosis and inflammation  occures in the course of neurodegenerative diseases, including Alzheimer's. It was also interesting that the celltype enrichment was found to be in the microglial cells, since these cells are of myeloid origin. That is, they come from macrophages, cells with mainly phagocytic and antigen-presenting activity, among others. Within the nervous tissue, they differentiate into microglia and fulfill immune functions, phagocyting foreign bodies such as apoptotic cells and amyloid deposits that are formed in the Alzheimer's disease. There are studies that relate microglial cells to the origin of the disease (Hemonnot _et al_. 2018).

The five genes with the highest MM of royalblue module are HYAL2, a liposomal hialuronidase enzyme, ICAM2, a intercellular binding protein, DGKH, a diacylglycerol kinase induced by glycocorticoids, CPD, a carboxypeptidase enzyme, and SNAT1, a transcriptional repressor. 
The most relevant functions in relation to the previously commented ones are that of ICAM2, which participates in the cell to cell union and interaction, something that forms part of the functions of microglial cells, and DGKH, because it regulates the levels of a second messenger such as diacylglycerol, which is involved in several signalling pathways.

### Tan

The Tan module contained genes that act as transcription factors. They were related with anabolic reactions and macromolecule biosynthesis. They were found to be associated with some signalling pathways such as MAP kinases and TFG (Tumor Growth Factor). One the one hand, there were studies that established a link between deficiency in TGF dependent signalling and Alzheimer's disease (Hu _et al_., 2019). On the other hand, there were genes associated with various types of cancer, such as lung, colorrectal, breast ant thyroid cancer; as well as the human leukemia virus, which it is capable to cause cancer due to its replication cycle. There are studies about an inverse correlation between cancer and Alzheimer's disease, since cancer is caused by an uncontrolled proliferation of cells and Alzheimer is caused by the opposite phenomenom, an increase in cell death (Anne Feng _et al_. , 2017). Again, the cell type associated with the genes of this module was the microglia.

The three genes with the highest MM are TBC1D10, an Rab type GTPase-activating protein involved in the regulation of vesicle transport in oligodendrocytes; OXSR1, a kinase that is activated in response to oxidative stress; and a non-traduced transcript, a non-coding RNA called LINC01354 with no attributed function.  Of these three, OXSR1 is perhaps the most relevant because of its relationship to oxidative stress and its role in apoptosis.

### Darkturquoise

The darkturquoise module had genes related to signalling pathways, immune response to virial infections through pathways such as interferon and proteins that interact with RNA. Its genes were expressed in the cellular cytosol. Cell types associated with this module were T helper lymphocytes and the microglia, both related with immune system. The fact that some  virus types may be related to the neuropathology of Alzheimer's disease is interesting, as stated in some studies (Sochocka _et al_ 2017).

CXCL5 is a cytokine from the CXC family. SAMD9L is a gen whose alteration is involved in diseases that affect the brain such as ataxia pancytopenia syndrome. UBA7 is involved in ubiquitination. PATL2 participates in RNA binding, and its alteration is related to fertility problems. 

### Purple

The purple module was enriched exclusively in mitochondria, with no associated cell types, focusing on metabolism and activity as a cofactor and oxidoreductase. Interestingly, it was involved in a multitude of metabolic processes, including basal metabolism and fatty acid metabolism. A certain association has been found between fatty acid metabolism and Alzheimer's disease (Snowden _et al_ 2017 ).

LINC00526 encodes for a long non-coding RNA. C1orf194 is a little studied gen and it would be the cause for different brain diseases. ZNF706 is a transcription factor involved in the negative regulation  of embryonic mesenchymal cell population maintenance, in addition to being involved in the negative regulation of transcription. RP4-612B15.3 is a pseudogen presents in immune system populations. TOE1 inhibits cell growth by stopping the cell cycle. 

### Floralwhite

The floralwhite module was related to the immune system at the level of intracellular (receptor signalling) and extracellular (immunoglobulin G (Ig G) ) signaling, as well as the activation of myeloid leukocytes. It has been found that during the development of Alzheimer's disease, there is expression of proinflammatory molecules in myeloid cells (Thome _et al_ 2018). In this module a relation was appreciated at the level of disease and cellular type, since it was enriched in rheumatoid arthritis and in differentiation of osteoclasts, and it had been seen that an excessive activity of the osteoclasts was related to the appearance of rheumatoid arthritis (Nakashima _et al_ 2008). No relationship was found between rheumatoid arthritis and Alzheimer's disease (Qixuan _et al_ 2018), so it would be possible that they only shared this pro-inflammatory profile and that is why it was highlighted in the analysis.
 
CPED1 is a gen of which there is no known function. PDCD4-AS1 encodes a long non-coding RNA. It is associated with breast cancer. C11orf45 encodes an uncharacterized protein. AC018462.3 encodes a long non-coding RNA. EMB is related with MAP kinase signalling pathway, and the transport of glucose and other sugar molecules. 



## Bibliography

Paula I. Moreira , Cristina Carvalho, Xiongwei Zhu,  Mark A. Smith , George Perry : Mitochondrial dysfunction is a trigger of Alzheimer's disease pathophysiology. _Biochimica et Biophysica_ Acta 2010, 1802 2:10

Anne-Laure Hemonnot , Jennifer Hua,  Lauriane Ulmann, and Hélène Hirbec Microglia in Alzheimer Disease: Well-Known Targets and New Opportunities. _Frontiers in Aging Neuroscience_ 2019, 11 233

Yueqiang Hu, Wei Chen, Lin Wu,  Lingfei Jiang , Ni Liang , Lulu Tan , Minghui Liang, Nong Tang: TGF-β1 Restores Hippocampal Synaptic Plasticity and Memory in Alzheimer Model via the PI3K/Akt/Wnt/β-Catenin Signaling Pathway. _Journal of Molecular Neuroscience_ 2019, 67, 142:149. 

Yen-Chen Anne Feng,  Kelly Cho, Sara Lindstrom,  Peter Kraft, Jean Cormack: Investigating the genetic relationship between Alzheimer’s disease and cancer using GWAS summary statistics. _Human Genetics_ 2017, 136, 1341:1351.

Marta Sochocka, Katarzyna Zwolińska, Jerzy Leszek. The Infectious Etiology of Alzheimer’s Disease. _Curr Neuropharmacol_ 2017, 15(7): 996–1009.

Stuart G. Snowden, Amera A. Ebshiana, Abdul Hye, Yang An, Olga Pletnikova, Richard O’Brien, John Troncoso, Cristina Legido-Quigley, Madhav Thambisetty. Association between fatty acid metabolism in the brain and Alzheimer disease neuropathology and cognitive performance: A nontargeted metabolomic study. _PLOS Medicine_ 2017, Mar 21;14(3):e1002266.

Aaron D. Thome, Alireza Faridar, David R. Beers, Jason R. Thonhoff, Weihua Zhao, Shixiang Wen, Belen Pascual, Joseph C. Masdeu,Stanley H. Appel. Functional alterations of myeloid cells during the course of Alzheimer’s disease.  _Molecular Neurodegeneration_ 2018, 13:61 

Qixuan Cai, Zhuoyuan Xin, Lin Zuo, Fan Li, Bin Liu. Alzheimer’s Disease and Rheumatoid Arthritis: A Mendelian Randomization Study.  _Frontiers in Neuroscience_ 2018, 12, 627.
